txp/Pos/StateLock/PrioLock.hs

```diff
@@ -0,0 +1,77 @@
+{-|
+Module:      Pos.StateLock.PrioLock
+Description: Provides a prioritised lock
+
+Provides a lock that can be taken with either high or low precedence.
+Within each precedence, the lock is taken in FIFO order.
+-}
+
+{-# LANGUAGE LambdaCase #-}
+
+module Pos.StateLock.PrioLock where
+
+import           Control.Concurrent.STM      (TMVar, newEmptyTMVar
+                                             , putTMVar, takeTMVar)
+import           Control.Monad.Catch         (MonadMask)
+import           Universum
+
+newtype PrioLock = PrioLock (TVar PrioLockState)
+
+data PrioLockState =
+    PUnlocked
+    | PLocked [TMVar ()] [TMVar ()] -- TODO: consider using queues instead of lists
```

Comment:
I had the silly "P" prefix just because originally it was in a module with a simple lock as well and I needed to avoid name clashes. We can drop the P prefix or suffix.

Reply:
Here's a simple Q impl:

```
data Q a = Q [a] [a]

queue :: [a] -> Q a
queue xs = Q xs []

enqueue :: Q a -> a -> Q a
enqueue (Q front  back) x = Q front (x : back)

dequeue :: Q a -> Maybe (a, Q a)
dequeue (Q (x:xs) back)    = Just (x, Q xs back)
dequeue (Q []     back)    = case reverse back of
                               x:xs -> Just (x, Q xs [])
                               []   -> Nothing
```
This impl is ok and is O(1) amortised time provided that the use of dequeue is non-persistent style, meaning that we never call dequeue twice on the same value. In our case that's true because it will be the unlocking thread that does the dequeue.

An alternative impl that would be suitable in other cases would do the reversing on the `enqueue` side.

Reply:
Thanks! I'll use that instead of the lists.